src/XrdVoms/XrdVomsMapfile.cc

```diff
@@ -420,51 +385,33 @@ XrdVomsMapfile::MaintenanceThread(void *myself_raw)
-       if (rval == -1) {
-           if (rval == EINTR) continue;
-           else break;
-       } else if (rval == 0) { // timeout!  Let's run maintenance.
-           next_update = monotonic_time_s() + m_update_interval;
-           struct stat statbuf;
-           if (-1 == stat(myself->m_mapfile.c_str(), &statbuf)) {
-               myself->m_edest->Emsg("XrdVomsMapfile", errno, "Error checking the mapfile",
-                   myself->m_mapfile.c_str());
-               myself->m_mapfile_ctime.tv_sec = 0;
-               myself->m_mapfile_ctime.tv_nsec = 0;
-               myself->m_is_valid = false;
-               continue;
-           }
-           if ((myself->m_mapfile_ctime.tv_sec == statbuf.st_ctim.tv_sec) &&
-               (myself->m_mapfile_ctime.tv_nsec == statbuf.st_ctim.tv_nsec))
-           {
-               myself->m_edest->Log(LogMask::Debug, "Maintenance", "Not reloading VOMS mapfile; "
-                   "no changes detected.");
-               continue;
-           }
-           memcpy(&myself->m_mapfile_ctime, &statbuf.st_ctim, sizeof(decltype(statbuf.st_ctim)));
-
-           myself->m_edest->Log(LogMask::Debug, "Maintenance", "Reloading VOMS mapfile now");
-           if ( !(myself->m_is_valid = myself->ParseMapfile(myself->m_mapfile)) ) {
-               myself->m_edest->Log(LogMask::Error, "Maintenance", "Failed to reload VOMS mapfile");
-           }
-       } else { // FD ready; let's shutdown
-           if (fds.revents & POLLIN) {
-               char indicator[1];
-               do {rval = read(myself->m_maintenance_pipe_r, indicator, 1);} while (rval != -1 || errno == EINTR);
-           }
+       auto rval = sleep(remaining);
+       if (rval > 0) {
+           // Woke up early due to a signal; re-run prior logic.
+           continue;
        }
-   }
-   if (errno) {
-       myself->m_edest->Emsg("Maintenance", "Failed to poll for events from parent object");
-   }
-   char indicator = '1';
-   int rval;
-   do {rval = write(myself->m_maintenance_thread_pipe_w, &indicator, 1);} while (rval != -1 || errno == EINTR);
-
+       next_update = monotonic_time_s() + m_update_interval;
+       struct stat statbuf;
+       if (-1 == stat(myself->m_mapfile.c_str(), &statbuf)) {
+           myself->m_edest->Emsg("XrdVomsMapfile", errno, "Error checking the mapfile",
+               myself->m_mapfile.c_str());
+           myself->m_mapfile_ctime.tv_sec = 0;
+           myself->m_mapfile_ctime.tv_nsec = 0;
+           myself->m_is_valid = false;
+           continue;
+       }
+       if ((myself->m_mapfile_ctime.tv_sec == statbuf.st_ctim.tv_sec) &&
+           (myself->m_mapfile_ctime.tv_nsec == statbuf.st_ctim.tv_nsec))
+       {
+           myself->m_edest->Log(LogMask::Debug, "Maintenance", "Not reloading VOMS mapfile; "
+               "no changes detected.");
+           continue;
+       }
+       memcpy(&myself->m_mapfile_ctime, &statbuf.st_ctim, sizeof(decltype(statbuf.st_ctim)));
 
-   return nullptr;
```

Comment:
The return nullptr line wants to stay:

/tmp/xrootd-5.4.1-rc1/src/XrdVoms/XrdVomsMapfile.cc: In static member function static void* XrdVomsMapfile::MaintenanceThread(void*)':
/tmp/xrootd-5.4.1-rc1/src/XrdVoms/XrdVomsMapfile.cc:423:1: error: no return statement in function returning non-void [-Werror=return-type]
 }
 ^
cc1plus: all warnings being treated as errors

Regards,
Giuseppe.